src/history/hs_rec.c

```diff
@@ -82,10 +82,16 @@ __hs_insert_record(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BTREE *btree,
     uint64_t upd_type_full_diag;
     int cmp;
 #endif
+    bool hs_read_all_flag;
     uint64_t counter, hs_counter;
     uint32_t hs_btree_id;
 
     counter = 0;
+    /*
+     * Keep track if the caller had set WT_CURSTD_HS_READ_ALL flag on the history store cursor. We
+     * want to preserve the flags set by the caller.
+     */
+    hs_read_all_flag = F_ISSET(cursor, WT_CURSTD_HS_READ_ALL);
```

Comment:
We need really a big comment about the way hs_read_all_flag is used.
Otherwise it is very confusing. Suggestion is, " the cursor functions should use the original hs_read_all flag from the caller, but search_near_after and search_near_before should be called by setting this flag always"

Reply:
Fixed in 63bf66b.